src/lib.rs

```diff
@@ -2729,6 +2731,68 @@ pub fn is_maximal_matching(
     })
 }
 
+/// Find the edges in the minimum spanning tree or forest of a graph
+/// using Kruskal's algorithm.
+///
+/// :param PyGraph graph: Undirected weighted graph with comparable weights
+///
+/// :returns: The :math:`N - |c|` edges of the Minimum Spanning Tree (or Forest, if :math:`|c| > 1`)
+///     where :math:`N` is the number of nodes and :math:`|c|` is the number of connected components of the graph
+/// :rtype: WeightedEdgeList
+#[pyfunction]
+#[text_signature = "(graph,)"]
+pub fn minimum_spanning_edges(
+    py: Python,
+    graph: &graph::PyGraph,
+) -> WeightedEdgeList {
+    let mut subgraphs = UnionFind::<usize>::new(graph.graph.node_bound());
+    let mut edge_list: Vec<EdgeReference<PyObject>> =
+        graph.graph.edge_references().collect();
+
+    edge_list.sort_by(|a, b| {
+        let weight_a = a.weight().as_ref(py);
+        let weight_b = b.weight().as_ref(py);
+        weight_a.compare(weight_b).unwrap()
+    });
+
+    let mut answer: Vec<(usize, usize, PyObject)> = Vec::new();
+    for edge in edge_list.iter() {
+        let u = edge.source().index();
+        let v = edge.target().index();
+        let w = edge.weight().clone();
+        if subgraphs.union(u, v) {
+            answer.push((u, v, w));
+        }
+    }
+
+    WeightedEdgeList { edges: answer }
+}
+
+/// Find the minimum spanning tree or forest of a graph
+/// using Kruskal's algorithm.
+///
+/// :param PyGraph graph: Undirected weighted graph with comparable weights
+///
+/// :returns: A Minimum Spanning Tree (or Forest, if the graph is not connected).
+///
+///     Note: The new graph will keep the same node indexes, but edge indexes might differ.
+/// :rtype: PyGraph
+#[pyfunction]
+#[text_signature = "(graph,)"]
+pub fn minimum_spanning_tree(
+    py: Python,
+    graph: &graph::PyGraph,
+) -> PyResult<graph::PyGraph> {
+    let mut spanning_tree = (*graph).clone();
```

Comment:
Is there any benefit of cloning the graph and clearing the edges over creating an empty graph and copying nodes from the original graph?

Reply:
Performance wise? No.

For keeping node indexes? It's a lot easier to code. `PyGraph` uses `StableGraph` as its base to store the graph, so there's no guarantee the node indexes are sequential. So we might have indexes `(0, 2, 4, 5)` in the old graph and when we insert it into the new graph they'll become sequential `(0, 1, 2, 3)` losing the mapping. Also, we cannot instantiate a graph with custom node indices because the variables from `petgraph` are private.

We can preserve this by looping `0 .. graph.node_bound()` and adding nodes to the new graph. Then we remove the indexes that are not in the old graph. And then we set the data to be the same as before. I found that code harder to implement and maintain

Reply:
Thanks! That makes sense, and I think the trade-off is reasonable here.